asmcomp/polling.ml

```diff
@@ -16,182 +17,226 @@ open Mach
+          end
+        in
+        let rec loop () =
+          while List.fold_left update false handlers do () done;
+          match List.find_opt (fun (n, _) -> not (get_handler n)) handlers with
+          | None -> ()
+          | Some (n, _) -> set_add_poll n; loop ()
+        in
+        loop ();
+        before body join exn
+    | Iexit n ->
+        get_handler n
+    | Itrywith(body, handler) ->
+        let join = before i.next end_ exn in
+        before body join (before handler join exn)
+    | Iraise _ ->
+        exn
+  in
+  let _r = before funbody true true in
+  assert _r;
+  get_add_poll
+
+(* Detection of functions that can loop via a tail-call without going
+   through a poll point. *)
+
+(* The result of the analysis is a single Boolean b.
+
+   b is true if there exists a path from the function entry to a
+   Potentially Recursive Tail Call (an Itailcall_ind or
+   Itailcall_imm to a forward function)
+   that does not go through an Ialloc or Ipoll instruction.
+
+   b is false, therefore, if the function always polls (via Ialloc or Ipoll)
+   before doing a PRTC.
+
+   To compute b, we do a backward dataflow analysis starting from
+   false, using || (Boolean "or") as the join operator, and with the
+   following transfer function:
+
+   TRANSF(Ialloc | Ipoll, x) = false
+   TRANSF(Itailcall_ind, x) = true
+   TRANSF(Itailcall_imm f, x) = f is a forward function
+   TRANSF(all other operations, x) = x
+*)
+
+let potentially_recursive_tailcall ~fwd_func funbody =
+  let handlers : (int, bool) Hashtbl.t = Hashtbl.create 20 in
+  let get_handler n =
+    match Hashtbl.find_opt handlers n with Some b -> b | None -> false in
+  let set_handler n b =
+    Hashtbl.replace handlers n b in
+  let rec before i end_ exn =
+    match i.desc with
+    | Iend -> end_
+    | Iop (Ialloc _ | Ipoll _) -> false
+    | Iop (Itailcall_ind) -> true
+    | Iop (Itailcall_imm { func }) -> String.Set.mem func fwd_func
+    | Iop op ->
+        if operation_can_raise op
+        then exn || before i.next end_ exn
+        else before i.next end_ exn
+    | Ireturn -> false
+    | Iifthenelse(_, ifso, ifnot) ->
+        let join = before i.next end_ exn in
+        before ifso join exn || before ifnot join exn
+    | Iswitch(_, branches) ->
+        let join = before i.next end_ exn in
+        Array.exists (fun i -> before i join exn) branches
+    | Icatch (Cmm.Nonrecursive, handlers, body) ->
+        let join = before i.next end_ exn in
+        List.iter
+          (fun (n, h) -> set_handler n (before h join exn))
+          handlers;
+        before body join exn
+    | Icatch (Cmm.Recursive, handlers, body) ->
+        let join = before i.next end_ exn in
+        let update changed (n, h) =
+          let b0 = get_handler n in
+          let b1 = before h join exn in
+          if b1 = b0 then changed else (set_handler n b1; true) in
+        while List.fold_left update false handlers do () done;
+        before body join exn
```

Comment:
I wonder if we really need a fixpoint iteration here.  We're looking for paths from entry to a PTRC that don't go through an Ialloc or Ipoll. If one such path exists, a loop-free path (one that does not go through a loop back-edge) exists too?